In [137]:
import pandas as pd
import numpy as np
from sklearn.cluster import KMeans 
import matplotlib.pyplot as plt
import folium
import matplotlib.cm as cm
import matplotlib.colors as colors

getting data out of wiki page

<h1>the first question 1</h1>

In [177]:
df = pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M", header=0)[0] 
df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

<h3>Delete rows with Not assigend Borough</h3>

In [178]:
df=df[df["Borough"] != "Not assigned"]
df.index = [x for x in range(df.shape[0])] 

In [179]:
df.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [162]:
df.shape

(103, 3)

In [163]:
df['Postal Code'].unique().shape

(103,)

the dataframe has 103 entries and the unique values in postal code are 103 that means there are no neighborhoods sharing the same postal code 

In [164]:
df1 = pd.read_csv('Geospatial_Coordinates .csv') #gettin postal codes data

In [165]:
ordered_postal_code = list(df["Postal Code"]) # getting a list of postal codes out of df

<h1>second question</h1>

In [166]:
df1 = df1.set_index("Postal Code").reindex(ordered_postal_code).reset_index()
# changing the order of the rows in df1 to match the same order of df according to postal code column

In [167]:
df1.head()

Postal Code   Latitude  Longitude
0         M3A  43.753259 -79.329656
1         M4A  43.725882 -79.315572
2         M5A  43.654260 -79.360636
3         M6A  43.718518 -79.464763
4         M7A  43.662301 -79.389494

In [168]:
lat_lan_cols = df1[['Latitude', 'Longitude']]

In [169]:
lat_lan_cols.head()

Latitude  Longitude
0  43.753259 -79.329656
1  43.725882 -79.315572
2  43.654260 -79.360636
3  43.718518 -79.464763
4  43.662301 -79.389494

In [170]:
df2 = pd.concat([df, lat_lan_cols], axis=1)

In [171]:
df2.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [172]:
df2_hotone = pd.get_dummies(df2)
df2_hotone.head()

Latitude  Longitude  Postal Code_M1B  Postal Code_M1C  Postal Code_M1E  \
0  43.753259 -79.329656                0                0                0   
1  43.725882 -79.315572                0                0                0   
2  43.654260 -79.360636                0                0                0   
3  43.718518 -79.464763                0                0                0   
4  43.662301 -79.389494                0                0                0   

   Postal Code_M1G  Postal Code_M1H  Postal Code_M1J  Postal Code_M1K  \
0                0                0                0                0   
1                0                0                0                0   
2                0                0                0                0   
3                0                0                0                0   
4                0                0                0                0   

   Postal Code_M1L  ...  Neighbourhood_Westmount  Neighbourhood_Weston  \
0                0  ...                        0                     0   
1                0  ...                        0                     0   
2                0  ...                        0                     0   
3                0  ...                        0                     0   
4                0  ...                        0                     0   

   Neighbourhood_Wexford, Maryvale  Neighbourhood_Willowdale, Newtonbrook  \
0                                0                                      0   
1                                0                                      0   
2                                0                                      0   
3                                0                                      0   
4                                0                                      0   

   Neighbourhood_Willowdale, Willowdale East  \
0                                          0   
1                                          0   
2                                          0   
3                                          0   
4                                          0   

   Neighbourhood_Willowdale, Willowdale West  Neighbourhood_Woburn  \
0                                          0                     0   
1                                          0                     0   
2                                          0                     0   
3                                          0                     0   
4                                          0                     0   

   Neighbourhood_Woodbine Heights  Neighbourhood_York Mills West  \
0                               0                              0   
1                               0                              0   
2                               0                              0   
3                               0                              0   
4                               0                              0   

   Neighbourhood_York Mills, Silver Hills  
0                                       0  
1                                       0  
2                                       0  
3                                       0  
4                                       0  

[5 rows x 214 columns]

In [173]:
kclusters = 5

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(df2_hotone)

kmeans.labels_[0:10] 

array([3, 3, 1, 3, 1, 0, 2, 3, 1, 1])

In [174]:
df2_hotone.insert(0, 'Cluster Labels', kmeans.labels_)
df2_hotone.head()

Cluster Labels   Latitude  Longitude  Postal Code_M1B  Postal Code_M1C  \
0               3  43.753259 -79.329656                0                0   
1               3  43.725882 -79.315572                0                0   
2               1  43.654260 -79.360636                0                0   
3               3  43.718518 -79.464763                0                0   
4               1  43.662301 -79.389494                0                0   

   Postal Code_M1E  Postal Code_M1G  Postal Code_M1H  Postal Code_M1J  \
0                0                0                0                0   
1                0                0                0                0   
2                0                0                0                0   
3                0                0                0                0   
4                0                0                0                0   

   Postal Code_M1K  ...  Neighbourhood_Westmount  Neighbourhood_Weston  \
0                0  ...                        0                     0   
1                0  ...                        0                     0   
2                0  ...                        0                     0   
3                0  ...                        0                     0   
4                0  ...                        0                     0   

   Neighbourhood_Wexford, Maryvale  Neighbourhood_Willowdale, Newtonbrook  \
0                                0                                      0   
1                                0                                      0   
2                                0                                      0   
3                                0                                      0   
4                                0                                      0   

   Neighbourhood_Willowdale, Willowdale East  \
0                                          0   
1                                          0   
2                                          0   
3                                          0   
4                                          0   

   Neighbourhood_Willowdale, Willowdale West  Neighbourhood_Woburn  \
0                                          0                     0   
1                                          0                     0   
2                                          0                     0   
3                                          0                     0   
4                                          0                     0   

   Neighbourhood_Woodbine Heights  Neighbourhood_York Mills West  \
0                               0                              0   
1                               0                              0   
2                               0                              0   
3                               0                              0   
4                               0                              0   

   Neighbourhood_York Mills, Silver Hills  
0                                       0  
1                                       0  
2                                       0  
3                                       0  
4                                       0  

[5 rows x 215 columns]

In [175]:
df2_hotone = pd.concat([df2_hotone, df2[["Neighbourhood"]]], axis=1)

In [176]:
df2_hotone.head()

Cluster Labels   Latitude  Longitude  Postal Code_M1B  Postal Code_M1C  \
0               3  43.753259 -79.329656                0                0   
1               3  43.725882 -79.315572                0                0   
2               1  43.654260 -79.360636                0                0   
3               3  43.718518 -79.464763                0                0   
4               1  43.662301 -79.389494                0                0   

   Postal Code_M1E  Postal Code_M1G  Postal Code_M1H  Postal Code_M1J  \
0                0                0                0                0   
1                0                0                0                0   
2                0                0                0                0   
3                0                0                0                0   
4                0                0                0                0   

   Postal Code_M1K  ...  Neighbourhood_Weston  \
0                0  ...                     0   
1                0  ...                     0   
2                0  ...                     0   
3                0  ...                     0   
4                0  ...                     0   

   Neighbourhood_Wexford, Maryvale  Neighbourhood_Willowdale, Newtonbrook  \
0                                0                                      0   
1                                0                                      0   
2                                0                                      0   
3                                0                                      0   
4                                0                                      0   

   Neighbourhood_Willowdale, Willowdale East  \
0                                          0   
1                                          0   
2                                          0   
3                                          0   
4                                          0   

   Neighbourhood_Willowdale, Willowdale West  Neighbourhood_Woburn  \
0                                          0                     0   
1                                          0                     0   
2                                          0                     0   
3                                          0                     0   
4                                          0                     0   

   Neighbourhood_Woodbine Heights  Neighbourhood_York Mills West  \
0                               0                              0   
1                               0                              0   
2                               0                              0   
3                               0                              0   
4                               0                              0   

   Neighbourhood_York Mills, Silver Hills  \
0                                       0   
1                                       0   
2                                       0   
3                                       0   
4                                       0   

                                 Neighbourhood  
0                                    Parkwoods  
1                             Victoria Village  
2                    Regent Park, Harbourfront  
3             Lawrence Manor, Lawrence Heights  
4  Queen's Park, Ontario Provincial Government  

[5 rows x 216 columns]

# third question

In [108]:
k_means_cluster_centers = kmeans.cluster_centers_
k_means_labels = kmeans.labels_

In [126]:
latitude, longitude = 43.733301, -79.372345 #coordinates of Toronto

In [135]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, name, cluster in zip(df2_hotone['Latitude'], df2_hotone['Longitude'], df2_hotone['Neighbourhood'], df2_hotone['Cluster Labels']):
    label = folium.Popup(str(name) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters